sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala

```diff
@@ -680,19 +681,63 @@ object JdbcUtils extends Logging {
   /**
    * Compute the schema string for this RDD.
    */
-  def schemaString(schema: StructType, url: String): String = {
+  def schemaString(
+      schema: StructType,
+      url: String,
+      createTableColumnTypes: Option[String] = None): String = {
     val sb = new StringBuilder()
     val dialect = JdbcDialects.get(url)
+    val userSpecifiedColTypesMap = createTableColumnTypes
+      .map(parseUserSpecifiedCreateTableColumnTypes(schema, _))
+      .getOrElse(Map.empty[String, String])
     schema.fields foreach { field =>
       val name = dialect.quoteIdentifier(field.name)
-      val typ: String = getJdbcType(field.dataType, dialect).databaseTypeDefinition
+      val typ: String = userSpecifiedColTypesMap.get(field.name)
+        .getOrElse(getJdbcType(field.dataType, dialect).databaseTypeDefinition)
       val nullable = if (field.nullable) "" else "NOT NULL"
       sb.append(s", $name $typ $nullable")
     }
     if (sb.length < 2) "" else sb.substring(2)
   }
 
   /**
+   * Parses the user specified createTableColumnTypes option value string specified in the same
+   * format as create table ddl column types, and returns Map of field name and the data type to
+   * use in-place of the default data type.
+   */
+  private def parseUserSpecifiedCreateTableColumnTypes(schema: StructType,
+    createTableColumnTypes: String): Map[String, String] = {
+    val userSchema = CatalystSqlParser.parseTableSchema(createTableColumnTypes)
+    val userColNames = userSchema.fieldNames
+    // check duplicate columns in the user specified column types.
+    if (userColNames.distinct.length != userColNames.length) {
+      val duplicates = userColNames.groupBy(identity).collect {
+        case (x, ys) if ys.length > 1 => x
+      }.mkString(", ")
+      throw new AnalysisException(
+        s"Found duplicate column(s) in createTableColumnTypes option value: $duplicates")
+    }
+    // check user specified column names exists in the data frame schema.
+    val commonNames = userColNames.intersect(schema.fieldNames)
```

Comment:
Thank you for the review. Good question., updated the PR with case-sensitive handling.  Now column names from user specified schema are matched with data frame schema based on the SQLConf.CASE_SENSITIVE flag.